io/disk/multipath_test.py

```diff
@@ -151,46 +151,161 @@ def test(self):
                 if disk in multipath.get_paths(path_dic["wwid"]):
                     msg += "Blacklist of %s fails\n" % disk
             multipath.form_conf_mpath_file(defaults_extra=plcy)
+        if msg:
+            self.fail("Some tests failed. Find details below:\n%s" % msg)
 
-            # Failing and reinstating individual paths eg: sdX
-            self.log.info(" Failing and reinstating the individual paths")
-            for path in path_dic["paths"]:
+    def test_fail_reinstate_individual_paths(self):
+        '''
+        Failing and reinstating individual paths eg: sdX
+        '''
+        err_paths = []
+        self.log.info(" Failing and reinstating the individual paths")
+        for wwid in self.wwids:
+            paths = multipath.get_paths(wwid)
```

Comment:
This is already done in line 104, why re-do ?

Reply:
The question applies to all other similar code as well.

Reply:
the paths we have in setup will have for only recent wwid and not for all wwid. I have tried this it doesn't run for all the wwids.
self.mpath_dic["paths"] = multipath.get_paths(wwid)
this will have only the last wwid paths.
by the way we have many library functions for multipaths now and we may need to remove this part. just checking wwid exist is enough and we can use the functions whenever we need.

Reply:
@narasimhan-v find the below job.log file for your reference.

Reply:
@Naresh-ibm my question is: when we have already found out mpaths in setUp, why are we finding it again in each test.

Reply:
I will re-iterate the same, the dictionary doesn't hold all the mpatha or paths for each wwid. it stores only the last wwid ran in that forloop.
hence to run the individual tes_functions for given number of wwids we have to use the for loop we need to roll them over for loop again.
hope you got it, if not we will discuss this over phone, to resolve this quickly and I can explain you in better way may be.

Reply:
dictionary holds only for last iteration, but it is a list of dictionaries. So, when you traverse the list and the mpath dictionary, it will provide you the whole set.

Reply:
Also, you are not traversing the older list, you are finding the mpaths each time, which is not ideal. That was my question about.

Reply:
[root@ltczz401-lp4 ~]# multipath -ll
Nov 21 04:31:09 | missing value for option 'wwid' on line 7 of /etc/multipath.conf
mpathc (360050768028383d7f00000000000011f) dm-5 IBM,2145
size=50G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=50 status=active
| |- 1:0:0:2 sdd 8:48  active ready running
| `- 2:0:1:2 sdm 8:192 active ready running
`-+- policy='service-time 0' prio=10 status=enabled
  |- 1:0:1:2 sdg 8:96  active ready running
  `- 2:0:0:2 sdj 8:144 active ready running
mpathb (360050768028383d7f00000000000011e) dm-4 IBM,2145
size=50G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=50 status=active
| |- 1:0:1:1 sdf 8:80  active ready running
| `- 2:0:0:1 sdi 8:128 active ready running
`-+- policy='service-time 0' prio=10 status=enabled
  |- 1:0:0:1 sdc 8:32  active ready running
  `- 2:0:1:1 sdl 8:176 active ready running
mpatha (360050768028383d7f00000000000011d) dm-3 IBM,2145
size=50G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=50 status=enabled
| |- 1:0:1:0 sde 8:64  active ready running
| `- 2:0:0:0 sdh 8:112 active ready running
`-+- policy='service-time 0' prio=10 status=enabled
  |- 1:0:0:0 sdb 8:16  active ready running
  `- 2:0:1:0 sdk 8:160 active ready running
[root@ltczz401-lp4 ~]# 



======>>>> I have passed "360050768028383d7f00000000000011d,360050768028383d7f00000000000011e"
360050768028383d7f00000000000011d(mpatha)----->>> sde, sdh, sdb, sdk
360050768028383d7f00000000000011d(mpathb)----->>> sdf, sdi, sdc, sdl

now dictionary holds on only 
360050768028383d7f00000000000011d(mpathb)----->>> sdf, sdi, sdc, sdl

and not the previos wwid values as it runs in for loops.

I agree that dictionary keys holds a list. but if you say dictionary looks as below,

dic[wwid]----> wwid1[sda, sdb, sdc], wwid2[sdd, sde,sdf]

then also we have to roll it over for loop which may take more number of steps then what we have now.

And As said in the begining we have the functions to do all this static parts, so we will remove them in future.
this is not going to require any more since we can handle using functions.

Reply:
@Naresh-ibm Definitely a discussion will help here, since we are on completely different pages on understanding.